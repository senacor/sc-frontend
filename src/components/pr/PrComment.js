import React from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles/index';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import TextField from '@material-ui/core/TextField';
import Collapse from '@material-ui/core/Collapse';
import MenuItem from '@material-ui/core/MenuItem';
import Icon from '@material-ui/core/Icon';
import Select from '@material-ui/core/Select';
import FormControl from '@material-ui/core/FormControl';
import Typography from '@material-ui/core/Typography';
import * as actions from '../../actions';
import PrSwipePositionDescription from './PrSwipePositionDescription';
import { debounce } from '../../helper/debounce';
import { isSupervisor } from '../../helper/checkRole';
import { translateContent } from '../translate/Translate';
import { getPrRatings, getUserroles } from '../../reducers/selector';

const styles = theme => ({
  nestedText: { paddingRight: '27px' },
  bootstrapInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '10px 12px',
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  containerListItem: {
    display: 'flex'
  },
  nestedListItem: {
    width: '80%'
  },
  iconComment: {
    color: theme.palette.primary['400']
  },
  iconNoComment: {
    color: '#dddddd'
  },
  nestedNumber: {
    width: '20%'
  },
  nestedTextSelect: {
    width: '95%'
  },
  comment: {
    paddingRight: '24px',
    color: theme.palette.primary['400'],
    fontStyle: 'italic'
  },
  expanded: {
    backgroundColor: 'rgba(0, 0, 0, 0.04)'
  },
  rating: {
    color: theme.palette.primary['400']
  }
});

// TODO Rename to PrReviewerRating
class PrComment extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      prById: this.props.prById,
      isExpanded: false
    };
  }
  handleChangeRating = (prById, category) => event => {
    this.setState({ [event.target.name]: event.target.value });

    this.props.prRating.rating = event.target.value;

    this.props.addRating(
      prById,
      category,
      this.props.prRating.comment,
      event.target.value,
      this.props.prRating.id
    );
  };
  handleChangeComment = (prById, category) => event => {
    this.setState({ comment: event.target.value });
    this.props.prRating.comment = event.target.value;

    this.sendComment(
      prById,
      category,
      event.target.value,
      this.state.rating,
      this.props.prRating.id
    );
  };
  sendComment = debounce(this.props.addRating, 500);
  handleClick = () => {
    this.setState({
      isExpanded: !this.state.isExpanded
    });
  };

  render() {
    const { prById, category, classes, prRating } = this.props;

    if (!prRating) {
      return null;
    }

    return (
      <div className={this.state.isExpanded ? classes.expanded : ''}>
        <div className={classes.containerListItem}>
          <ListItem
            button
            className={classes.nestedListItem}
            onClick={() => this.handleClick(category)}
          >
            <ListItemText
              id={`${category}_CommentDescription`}
              secondary={translateContent(category)}
            />

            <Icon
              id={`${category}_CommentIconId`}
              className={
                prRating.comment && this.props.prVisible
                  ? classes.iconComment
                  : classes.iconNoComment
              }
            >
              comment
            </Icon>
          </ListItem>

          <ListItem className={classes.nestedNumber}>
            {!isSupervisor(this.props.userroles) ? (
              <Typography
                id={category}
                className={classes.rating}
                variant="body2"
              >
                {this.props.prVisible ? prRating.rating : ''}
              </Typography>
            ) : (
              <FormControl
                className={classes.formControl}
                disabled={this.props.prFinalized}
              >
                <Select
                  id={category + '_RatingId'}
                  value={prRating.rating ? prRating.rating : 3}
                  onChange={this.handleChangeRating(prById, category)}
                  displayEmpty
                  name="rating"
                >
                  {[1, 2, 3, 4, 5].map(ratingValue => {
                    return (
                      <MenuItem
                        key={category + '_RatingValue' + ratingValue}
                        id={category + '_RatingValue' + ratingValue}
                        value={ratingValue}
                      >
                        {ratingValue}
                      </MenuItem>
                    );
                  })}
                </Select>
              </FormControl>
            )}
          </ListItem>
        </div>
        <Collapse in={this.state.isExpanded} timeout="auto" unmountOnExit>
          <List component="div" disablePadding className={classes.nestedText}>
            {!isSupervisor(this.props.userroles) ? (
              <ListItem>
                <Typography
                  id={category + '_TYPO'}
                  className={classes.comment}
                  variant="body1"
                >
                  {prRating.comment && this.props.prVisible
                    ? '»' + prRating.comment + '«'
                    : 'Noch nicht freigegeben.'}
                </Typography>
              </ListItem>
            ) : (
              <ListItem>
                <TextField
                  id={category + '_CommentId'}
                  label="Kommentar"
                  multiline
                  fullWidth
                  rowsMax="4"
                  value={
                    this.state.comment ? this.state.comment : prRating.comment
                  }
                  disabled={this.props.prFinalized}
                  onChange={this.handleChangeComment(prById, category)}
                  InputProps={{
                    disableUnderline: true,
                    name: 'comment',
                    classes: {
                      input: classes.bootstrapInput
                    }
                  }}
                  InputLabelProps={{
                    shrink: true
                  }}
                />
              </ListItem>
            )}
            <PrSwipePositionDescription category={category} />
          </List>
        </Collapse>
      </div>
    );
  }
}

export const StyledComponent = withStyles(styles)(PrComment);
export default connect(
  (state, props) => ({
    userroles: getUserroles(state),
    prRating: getPrRatings(props.category)(state)
  }),
  {
    addRating: actions.addRating
  }
)(StyledComponent);
