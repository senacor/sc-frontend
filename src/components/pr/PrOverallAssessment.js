import React from 'react';
import { withStyles } from '@material-ui/core/styles/index';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import TextField from '@material-ui/core/TextField';
import { connect } from 'react-redux';
import MenuItem from '@material-ui/core/MenuItem';
import Icon from '@material-ui/core/Icon';
import Select from '@material-ui/core/Select';
import FormControl from '@material-ui/core/FormControl';
import * as actions from '../../actions';
import { debounce } from '../../helper/debounce';

const styles = theme => ({
  nestedText: {
    paddingLeft: '20px',
    paddingRight: '27px'
  },
  bootstrapInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '10px 12px',
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  containerListItem: {
    display: 'flex'
  },
  description: {
    paddingLeft: '0',
    paddingRight: '30px'
  },
  nestedListItem: {
    paddingLeft: '20px',
    width: '80%'
  },
  nestedListItemFTF: {
    maxWidth: '80%'
  },
  icon: {
    color: theme.palette.primary['400']
  },
  iconWhite: {
    color: theme.palette.common.white
  },
  nestedNumber: {
    minWidth: '44%',
    justifyContent: 'flex-end'
  },
  nestedNumberTargetRole: {
    width: '95%'
  },
  formControl: {
    paddingRight: '0px',
    minWidth: '95%'
  }
});

class PrOverallAssessment extends React.Component {
  constructor(props) {
    super(props);
    const categoryFulfillment = 'FULFILLMENT_OF_REQUIREMENT';
    const categoryTargetRole = 'TARGET_ROLE';

    this.state = {
      ratingFulfillment: this.props.prById.prRatingSet.find(
        prRating => prRating.prRatingDescription === categoryFulfillment
      ).rating,
      ratingTargetRole: this.props.prById.prRatingSet.find(
        prRating => prRating.prRatingDescription === categoryTargetRole
      ).rating,
      comment: this.props.prById.prRatingSet.find(
        prRating => prRating.prRatingDescription === categoryFulfillment
      ).comment,
      prById: this.props.prById,
      categoryFulfillment: categoryFulfillment,
      categoryTargetRole: categoryTargetRole,
      is_expanded: false
    };
  }

  handleChangeRating = prById => event => {
    this.setState({ [event.target.name]: event.target.value });
    let category = (event.target.name = 'ratingFulfillment'
      ? this.state.categoryFulfillment
      : this.state.categoryTargetRole);
    let prRating = prById.prRatingSet.find(
      prRating => prRating.prRatingDescription === category
    );
    this.props.addRating(
      prById,
      category,
      this.state.comment,
      event.target.value,
      prRating.id
    );
  };

  handleChangeComment = prById => event => {
    this.setState({ [event.target.name]: event.target.value });
    let prRating = prById.prRatingSet.find(
      prRating =>
        prRating.prRatingDescription === this.state.categoryFulfillment
    );

    this.sendComment(
      prById,
      this.state.categoryFulfillment,
      event.target.value,
      this.state.ratingFulfillment,
      prRating.id
    );
  };

  sendComment = debounce(this.props.addRating, 500);

  translateRatingDescription = ratingDescription => {
    switch (ratingDescription) {
      case 'FULFILLMENT_OF_REQUIREMENT':
        return 'Gesamteinschätzung; Erfüllung der Anforderungen für aktuelle Stufe';
      case 'TARGET_ROLE':
        return 'Zielrolle (ab Senior Level ausfüllen)';
      default:
        return 'default';
    }
  };

  categoryText = {
    text:
      'In welchem Umfang erfüllt die Mitarbeiterin/der Mitarbeiter die Anforderungen an seine aktuelle Laufbahnstufe vor dem Hintergrund der aktuellen Einstufung? Welche Stärken gilt es auszubauen, welche Lücken sollten geschlossen werden?'
  };

  render() {
    const { prById, classes } = this.props;

    return (
      <div>
        <div className={classes.containerListItem}>
          <ListItem>
            <ListItemText
              secondary={this.translateRatingDescription(
                this.state.categoryFulfillment
              )}
            />

            {this.state.comment ? (
              <Icon className={classes.icon}>comment</Icon>
            ) : (
              <Icon className={classes.iconWhite}>comment</Icon>
            )}
          </ListItem>

          <ListItem className={classes.nestedNumber}>
            <FormControl className={classes.formControl}>
              <Select
                value={
                  this.state.ratingFulfillment
                    ? this.state.ratingFulfillment
                    : 3
                }
                onChange={this.handleChangeRating(prById)}
                displayEmpty
                name="ratingFulfillment"
              >
                <MenuItem value={1}>nicht erfüllt</MenuItem>,
                <MenuItem value={2}>zT. n. erfüllt</MenuItem>,
                <MenuItem value={3}>erfüllt</MenuItem>,
                <MenuItem value={4}>zT. übererfüllt</MenuItem>,
                <MenuItem value={5}>übererfüllt</MenuItem>
              </Select>
            </FormControl>
          </ListItem>
        </div>
        <List component="div" disablePadding className={classes.nestedText}>
          <ListItem>
            <ListItemText
              secondary={this.translateRatingDescription(
                this.state.categoryTargetRole
              )}
            />
            <FormControl className={classes.nestedNumberTargetRole}>
              <Select
                value={
                  this.state.ratingTargetRole ? this.state.ratingTargetRole : 1
                }
                onChange={this.handleChangeRating(prById)}
                displayEmpty
                name="ratingTargetRole"
              >
                <MenuItem value={1}>keine Auswahl</MenuItem>,
                <MenuItem value={2}>Platformgestalter</MenuItem>,
                <MenuItem value={3}>Business IT Solution Leader</MenuItem>,
                <MenuItem value={4}>Transformation Manager</MenuItem>,
                <MenuItem value={5}>IT-Liefersteuerer</MenuItem>,
                <MenuItem value={6}>Architect</MenuItem>,
                <MenuItem value={7}>Technical Expert</MenuItem>,
                <MenuItem value={8}>Lead Developer</MenuItem>
              </Select>
            </FormControl>
          </ListItem>
          <ListItem>
            <TextField
              id={this.state.categoryFulfillment}
              label="Freitextfeld (bitte ausfüllen)"
              multiline
              fullWidth
              rowsMax="4"
              value={this.state.comment ? this.state.comment : ''}
              onChange={this.handleChangeComment(prById)}
              InputProps={{
                disableUnderline: true,
                name: 'comment',
                classes: {
                  input: classes.bootstrapInput
                }
              }}
              InputLabelProps={{
                shrink: true
              }}
            />
          </ListItem>
          <ListItem />
          <ListItem>
            <ListItemText
              className={classes.description}
              primary="Beschreibung"
              secondary={this.categoryText.text}
            />
          </ListItem>
        </List>
      </div>
    );
  }
}

export const StyledComponentOA = withStyles(styles)(PrOverallAssessment);
export default connect(
  state => ({
    prRating: state.prRatings.prRating
  }),
  {
    addRating: actions.addRating
  }
)(StyledComponentOA);
