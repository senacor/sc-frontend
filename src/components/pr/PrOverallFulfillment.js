import React from 'react';
import { connect } from 'react-redux';

import ListItem from '@material-ui/core/ListItem';
import FormControl from '@material-ui/core/FormControl';
import Typography from '@material-ui/core/Typography';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';

import { getPrRatings } from '../../reducers/selector';
import * as actions from '../../actions';
import { withStyles } from '@material-ui/core';
import { mapRatingFullfilment } from '../../helper/mapRatingFullfilment';
import { injectIntl } from 'react-intl';

const styles = {
  simpleBlack: {
    color: '#000000',
    width: '80%'
  },
  number: {
    width: '20%',
    paddingLeft: '5%'
  }
};

const PrOverallFulfillment = ({
  category,
  prRating,
  prById,
  addRating,
  readOnly,
  isActionPerformer,
  nonActionPerformer,
  openEditing,
  classes,
  intl
}) => {
  if (!prRating) {
    return null;
  }

  const handleChangeRating = event => {
    addRating(
      prById,
      category,
      prRating.comment,
      event.target.value,
      prRating.id
    );
  };

  if (nonActionPerformer) {
    return (
      <ListItem>
        <div className={classes.simpleBlack}>
          <Typography>
            {intl.formatMessage({
              id: `${category}`
            })}
          </Typography>
        </div>
        <Typography id="FULFILLMENT_OF_REQUIREMENT_TYPO" variant="body1">
          {readOnly
            ? mapRatingFullfilment(prRating.rating, intl)
            : intl.formatMessage({
                id: 'proverallfulfillment.noentry'
              })}
        </Typography>
      </ListItem>
    );
  } else if (isActionPerformer) {
    return (
      <ListItem>
        <div className={classes.simpleBlack}>
          <Typography>
            {intl.formatMessage({
              id: `${category}`
            })}
          </Typography>
        </div>
        <div className={classes.number}>
          <FormControl disabled={!openEditing}>
            <Select
              id="FULFILLMENT_OF_REQUIREMENT_RatingId"
              value={prRating.rating ? prRating.rating : 0}
              onChange={handleChangeRating}
              displayEmpty
              name="ratingFulfillment"
            >
              {[0, 1, 2, 3, 4, 5].map(ratingValue => {
                return (
                  <MenuItem
                    key={
                      'FULFILLMENT_OF_REQUIREMENT' +
                      '_RatingValue' +
                      ratingValue
                    }
                    id={
                      'FULFILLMENT_OF_REQUIREMENT' +
                      '_RatingValue' +
                      ratingValue
                    }
                    value={ratingValue}
                  >
                    {mapRatingFullfilment(ratingValue, intl)}
                  </MenuItem>
                );
              })}
            </Select>
          </FormControl>
        </div>
      </ListItem>
    );
  } else {
    return null;
  }
};

export const StyledComponent = withStyles(styles)(PrOverallFulfillment);
export default injectIntl(
  connect(
    (state, props) => ({
      prRating: getPrRatings(props.category)(state)
    }),
    {
      addRating: actions.addRating
    }
  )(StyledComponent)
);
