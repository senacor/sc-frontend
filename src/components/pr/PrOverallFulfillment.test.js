import React from 'react';
import { shallow } from 'enzyme';
import { PrOverallFulfillment } from './PrOverallFulfillment';
import { StyledComponent } from './PrComment';
import ROLES from '../../helper/roles';
import FormControl from '@material-ui/core/FormControl';
import TextField from '@material-ui/core/TextField';
import { createShallow } from '@material-ui/core/test-utils';
import { StyledComponentOA } from './PrOverallAssessment';

describe('PrOverallFulfillment Component', () => {
  let shallow = createShallow({ dive: true });

  it('should match snapshot', () => {
    let cut = shallow(
      <PrOverallFulfillment
        prById={{}}
        prRating={{
          id: 5,
          prRatingDescription: 'FULFILLMENT_OF_REQUIREMENT',
          prRatingCategory: 'PERFORMANCE_IN_PROJECT',
          rating: 2,
          comment: 'ggg'
        }}
        category="TARGET_ROLE"
      />
    );

    expect(cut).toMatchSnapshot();
  });

  it('should display the fulfillment of requirement', () => {
    let cut = shallow(
      <PrOverallFulfillment
        prById={{}}
        prRating={{
          id: 5,
          prRatingDescription: 'FULFILLMENT_OF_REQUIREMENT',
          prRatingCategory: 'PERFORMANCE_IN_PROJECT',
          rating: 2,
          comment: 'ggg'
        }}
        category="TARGET_ROLE"
      />
    );

    expect(cut).toMatchSnapshot();
  });

  it('should display the fulfillment of requirement as read-only if the supervisor has finalized the PR', () => {
    const wrapper = shallow(
      <PrOverallFulfillment
        prById={{}}
        prRating={{
          id: 5,
          prRatingDescription: 'FULFILLMENT_OF_REQUIREMENT',
          prRatingCategory: 'PERFORMANCE_IN_PROJECT',
          rating: 2,
          comment: 'ggg'
        }}
        category="TARGET_ROLE"
        prFinalized={true}
      />
    );

    expect(
      wrapper
        .find(FormControl)
        .find('[disabled]')
        .props().disabled
    ).toEqual(true);
  });
});
