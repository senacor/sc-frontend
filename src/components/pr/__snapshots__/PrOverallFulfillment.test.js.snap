// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrOverallFulfillment Component should display the fulfillment of requirement 1`] = `
<WithStyles(ListItem)>
  <div
    className="PrOverallFulfillment-simpleBlack-1"
  >
    <WithStyles(Typography)>
      Zielrolle (ab Senior Level ausfüllen)
    </WithStyles(Typography)>
  </div>
  <WithStyles(Typography)
    id="FULFILLMENT_OF_REQUIREMENT_TYPO"
    variant="body1"
  >
    kein Eintrag
  </WithStyles(Typography)>
</WithStyles(ListItem)>
`;

exports[`PrOverallFulfillment Component should match snapshot 1`] = `
<WithStyles(ListItem)>
  <div
    className="PrOverallFulfillment-simpleBlack-1"
  >
    <WithStyles(Typography)>
      Zielrolle (ab Senior Level ausfüllen)
    </WithStyles(Typography)>
  </div>
  <div
    className="PrOverallFulfillment-number-2"
  >
    <WithStyles(FormControl)>
      <WithStyles(Select)
        displayEmpty={true}
        id="ratingFullfillmentId"
        name="ratingFulfillment"
        onChange={[Function]}
        value={2}
      >
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue0"
          key="_ratingFulfillment0"
          value={0}
        >
          kein Eintrag
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue1"
          key="_ratingFulfillment1"
          value={1}
        >
          nicht erfüllt
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue2"
          key="_ratingFulfillment2"
          value={2}
        >
          zT. nicht erfüllt
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue3"
          key="_ratingFulfillment3"
          value={3}
        >
          erfüllt
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue4"
          key="_ratingFulfillment4"
          value={4}
        >
          zT. übererfüllt
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          id="_ratingFulfillmentValue5"
          key="_ratingFulfillment5"
          value={5}
        >
          übererfüllt
        </WithStyles(MenuItem)>
      </WithStyles(Select)>
    </WithStyles(FormControl)>
  </div>
</WithStyles(ListItem)>
`;
